lib/events/emitter_test.go

```diff
@@ -129,3 +131,109 @@ func TestWriterEmitter(t *testing.T) {
 		assert.Contains(t, scanner.Text(), events[i].GetCode())
 	}
 }
+
+func TestAsyncEmitter(t *testing.T) {
+	clock := clockwork.NewRealClock()
+	events := GenerateTestSession(SessionParams{PrintEvents: 20})
+
+	// Slow tests that async emitter does not block
+	// on slow emitters
+	t.Run("Slow", func(t *testing.T) {
+		emitter, err := NewAsyncEmitter(AsyncEmitterConfig{
+			Inner: &slowEmitter{clock: clock, timeout: time.Hour},
+		})
+		assert.NoError(t, err)
+		defer emitter.Close()
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+		defer cancel()
+		err = emitter.EmitAuditEvent(ctx, events[0])
+		assert.NoError(t, err)
+		assert.NoError(t, ctx.Err())
+	})
+
+	// Receive makes sure all events are recevied in the same order as they are sent
+	t.Run("Receive", func(t *testing.T) {
+		chanEmitter := &channelEmitter{eventsCh: make(chan AuditEvent, len(events))}
+		emitter, err := NewAsyncEmitter(AsyncEmitterConfig{
+			Inner: chanEmitter,
+		})
+
+		assert.NoError(t, err)
+		defer emitter.Close()
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+		defer cancel()
+		for _, event := range events {
+			err := emitter.EmitAuditEvent(ctx, event)
+			assert.NoError(t, err)
+		}
+
+		for i := 0; i < len(events); i++ {
+			select {
+			case event := <-chanEmitter.eventsCh:
+				assert.Equal(t, events[i], event)
+			case <-time.After(time.Second):
+				t.Fatalf("timeout")
+			}
+		}
+	})
+
+	// Close makes sure that close cancels operations and context
+	t.Run("Close", func(t *testing.T) {
+		counter := &counterEmitter{}
+		emitter, err := NewAsyncEmitter(AsyncEmitterConfig{
+			Inner:      counter,
+			BufferSize: len(events),
+		})
+		assert.NoError(t, err)
+
+		ctx, cancel := context.WithTimeout(context.TODO(), time.Second)
+		defer cancel()
+
+		for i := 0; i < len(events); i++ {
+			go emitter.EmitAuditEvent(ctx, events[i])
+		}
+
+		// context will not wait until all events have been submitted
+		emitter.Close()
+		assert.True(t, int(counter.count.Load()) <= len(events))
+
+		// make sure context is done to prevent context leaks
+		select {
+		case <-emitter.ctx.Done():
+		default:
+			t.Fatalf("Context leak, should be closed")
+		}
+	})
+}
+
+type slowEmitter struct {
+	clock   clockwork.Clock
```

Comment:
No need for the clockwork's clock here since the test explicitly blocking - the implementation might be simpler with the channel:
```go
type emitter struct {
 	c <-chan time.Time
}
```
Also, another simplification could be the use of an emitterFunc that would trivially implement Emitter with itself:
```go
type func emitterFunc(context.Context, AuditEvent)
```
which then could be used to implement simple emitters:
```go
emitter, err := NewAsyncEmitter(AsyncEmitterConfig{
	Inner: emitterFunc(func(context.Context, AuditEvent) {<-time.After(time.Hour)})
})
```

Reply:
Clockwork allows to use real time and fake time with other helpful methods, I would prefer to use it in tests when possible.